libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs

```diff
@@ -0,0 +1,95 @@
+                Speak = Speak ?? defaultPromptOptions.Speak;
+                RetryPromptString = RetryPromptString ?? defaultPromptOptions.RetryPromptString;
+                RetryPromptActivity = RetryPromptActivity ?? defaultPromptOptions.RetryPromptActivity;
+                RetrySpeak = RetrySpeak ?? defaultPromptOptions.RetrySpeak;
+            }
+        }
+
+        public DateTime Expires { get; set; }
+    }
+
+    public class OAuthPrompt : Control, IDialog, IDialogContinue
+    {
+        private prompts.OAuthPrompt _prompt;
+        private OAuthPromptSettingsWithTimeout _settings;
+
+        public OAuthPrompt(OAuthPromptSettingsWithTimeout settings, PromptValidator<TokenResult> validator = null)
+        {
+            _settings = settings ?? throw new ArgumentException(nameof(settings));
+            _prompt = new prompts.OAuthPrompt(settings, validator);
+        }
+
+        public async Task DialogBegin(DialogContext dc, object dialogArgs = null)
+        {
+            if (dc == null)
+                throw new ArgumentNullException(nameof(dc));
+            if (dialogArgs == null)
+                throw new ArgumentNullException(nameof(dialogArgs));
+
+            var promptOptions = (PromptOptions)dialogArgs;
+
+            //persist options and state
+            var timeout = _settings.Timeout.HasValue ? _settings.Timeout.Value : 54000000;
+            var instance = dc.Instance;
+            instance.State = new OAuthPromptState(promptOptions);
+
+            var tokenResult = await _prompt.GetUserToken(dc.Context);
+
+            if (tokenResult != null && tokenResult.Value != null)
+            {
+                await dc.End(tokenResult);
+            }
+            else if (!string.IsNullOrEmpty(promptOptions.PromptString))
+            {
+                await _prompt.Prompt(dc.Context, promptOptions.PromptString, promptOptions.Speak);
+            }
+            else if (promptOptions.PromptActivity != null)
+            {
+                await _prompt.Prompt(dc.Context, promptOptions.PromptActivity);
+            }
+        }
+
+        public async Task DialogContinue(DialogContext dc)
+        {
+            if (dc == null)
+                throw new ArgumentNullException(nameof(dc));
+            //Recognize token
+            var tokenResult = await _prompt.Recognize(dc.Context);
+            //Check for timeout
+            var state = dc.Instance.State as OAuthPromptState;
+            var isMessage = dc.Context.Activity.Type == ActivityTypes.Message;
+            var hasTimedOut = isMessage && (DateTime.Compare(DateTime.Now, state.Expires) > 0);
+
+            if (tokenResult == null || hasTimedOut)
+            {
+                await dc.End(tokenResult);
+            }
+            else if (isMessage && !string.IsNullOrEmpty(state.RetryPromptString))
+            {
+                await dc.Context.SendActivity(state.RetryPromptString, state.RetrySpeak);
+            }
```

Comment:
There seem to be a number of states through here that don't hit any code. 

For example: 
1. activityType != message doesn't seem to do anything here. 
2. Many other states. 

Please clean up the code so that it's much more clear what the intent is.

Reply:
this is intentional. at this state, if we weren't able to fetch a token for the user, then either the credentials were invalid or there is a retry scheduled to try to get the credentials again. if neighter is possible, then the prompt just ends with a null tokenresult. so checking for other states isnt necessary

---
In reply to: [184819720](https://github.com/Microsoft/botbuilder-dotnet/pull/520#discussion_r184819720) [](ancestors = 184819720)